### PR TITLE
WIP: Clean up service resources when stopWithApp is true

### LINKED ISSUE #732: URGENT: Trackplayer is not removed from memory when app closes
Using trackplayer 1.20.

I see the app stays in memory even when using StopWithApp = true.
It looks like there are two instances fired and one is going down after app and other stays in memory. 

Also i can confirm that the app stays in the background as i get reports data traffic is out in 4 days.

Where do i need to look to solve this?

My part of loading: (maybe buggy haha)

```
import React, { Component, useEffect, useState } from 'react';
import {
  Platform,
  StyleSheet,
  Text,
  View,
  Image,
  TouchableOpacity,
} from 'react-native';
import { connect } from 'react-redux';
import { togglePlayer } from '../actions';
import NetInfo from '@react-native-community/netinfo';

import TrackPlayer, { usePlaybackState } from 'react-native-track-player';

import * as playerSetting from '../constants/PlayerConstants';
import * as screenSetting from '../constants/ScreenConstants';

const { playerUrl } = playerSetting;
const { width, height } = screenSetting;

function Player(props) {
  const playbackState = usePlaybackState();
  const [haveInternet, setHaveInternet] = useState(true);

  useEffect(
    () => {
      //Event Listeners
      NetInfo.addEventListener(
        'connectionChange',
        handleFirstConnectivityChange
      );
      function handleFirstConnectivityChange() {
        TrackPlayer.reset();
        if (Platform.OS === 'ios') {
          togglePlayback();
        }
      }

      NetInfo.isConnected.addEventListener(
        'connectionChange',
        this.handleConnectivityChange
      );

      NetInfo.isConnected.fetch().done((isConnected) => {
        if (isConnected == true) {
          setHaveInternet(true);
        }
      });
      TrackPlayer.setupPlayer({});
      TrackPlayer.updateOptions({
        stopWithApp: true,
        //waitForBuffer: true,
        capabilities: [
          TrackPlayer.CAPABILITY_PLAY,
          TrackPlayer.CAPABILITY_PAUSE,
          TrackPlayer.CAPABILITY_STOP,
        ],
        compactCapabilities: [
          TrackPlayer.CAPABILITY_PLAY,
          TrackPlayer.CAPABILITY_PAUSE,
        ],
      });

      return () => {
        //TrackPlayer.destroy();
        NetInfo.isConnected.removeEventListener(
          'connectionChange',
          this.handleConnectivityChange
        );
        NetInfo.removeEventListener(
          'connectionChange',
          handleFirstConnectivityChange
        );
      };
    },
    [TrackPlayer.stop(), TrackPlayer.reset()]
  );

  handleConnectivityChange = (isConnected) => {
    if (isConnected == false) {
      setHaveInternet(false);
      //togglePlayback();
    } else {
      setHaveInternet(true);
      //togglePlayback();
    }
  };

  renderButton = () => {
    if (
      playbackState == TrackPlayer.STATE_PAUSED ||
      playbackState === TrackPlayer.STATE_STOPPED ||
      playbackState === TrackPlayer.STATE_NONE
    ) {
      return (
        <Image
          source={require('../assets/play.png')}
          style={styles.playerButton}
          imageStyle={{ resizeMode: 'contain' }}
        />
      );
    } else {
      return (
        <Image
          source={require('../assets/pauze.png')}
          style={styles.playerButton}
          imageStyle={{ resizeMode: 'contain' }}
        />
      );
    }
  };

  async function togglePlayback() {
    const currentTrack = await TrackPlayer.getCurrentTrack();
    if (currentTrack == null) {
      await TrackPlayer.add({
        id: 'HeteHits',
        url: playerUrl,
        title: 'Loading...',
        artist: 'Loading...',
        artwork: '../assets/track_image.png',
      });
      await TrackPlayer.play();
      props.togglePlayerState();
    } else {
      if (playbackState === TrackPlayer.STATE_PAUSED) {
        await TrackPlayer.play();
        props.togglePlayerState();
      } else {
        TrackPlayer.pause();
        props.togglePlayerState();
      }
    }
  }

  return (
    <View style={styles.container}>
      <TouchableOpacity
        disabled={!haveInternet}
        onPress={() => togglePlayback()}
      >
        <View
          style={{
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'space-between',
          }}
        >
          {renderButton()}
          <Image
            source={require(`../assets/luisterLive.png`)}
            style={{
              height: height * 0.075,
              width: width * 0.175,
              resizeMode: 'contain',
            }}
            imageStyle={{ resizeMode: 'contain' }}
          />
        </View>
      </TouchableOpacity>
    </View>
  );
}

const mapStateToProps = (state) => {
  return {
    speeltNu: state.togglePlayer.speeltNu,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    togglePlayerState: () => dispatch(togglePlayer()),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Player);

const styles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  playerButton: {
    height: height * 0.07,
    width: width * 0.17,
    resizeMode: 'contain',
    marginBottom: 10,
  },
});

```

### PR DESCRIPTION
**Pull request by skahack**
Fixes #732

### CODE REVIEW

File: android/src/main/java/com/guichaguri/trackplayer/service/MusicService.java [hunk 1]
```diff
@@ -127,6 +127,8 @@ public void onTaskRemoved(Intent rootIntent) {
         super.onTaskRemoved(rootIntent);
 
         if (manager == null || manager.shouldStopWithApp()) {
+            manager.getPlayback().stop();
+            destroy();
```
Comment (Guichaguri) [inline android/src/main/java/com/guichaguri/trackplayer/service/MusicService.java]: I don't see how this fixes the problem. `stopSelf()` should automatically call `onDestroy`, which will cleanup the service resources.

Comment (skahack) [reply android/src/main/java/com/guichaguri/trackplayer/service/MusicService.java]: I investigated but onDestory was not called. I don't know where this behavior is documented.

related?: https://stackoverflow.com/questions/7236782/is-an-android-service-guaranteed-to-call-ondestroy/7237522#7237522

Comment (skahack) [reply android/src/main/java/com/guichaguri/trackplayer/service/MusicService.java]: I'm sorry. It was a misunderstanding.
OnDestroy is called, but I don't know if it's the effect of the release order, but the app state remains in memory.

**Comment by curiousdustin** [comment]
App state living on is the same issue I think I’m having in #813

File: android/src/main/java/com/guichaguri/trackplayer/service/MusicService.java (hunk 1 above)
Comment (illl48) [reply android/src/main/java/com/guichaguri/trackplayer/service/MusicService.java]: do we need to check if manager == null  before callin manager.getPlayback().stop();?

Comment (skahack) [reply android/src/main/java/com/guichaguri/trackplayer/service/MusicService.java]: You are right. Thanks.

**Comment by Guichaguri** [comment]
@curiousdustin have you tested this PR? does it fixes your issue?

**Comment by curiousdustin** [comment]
@Guichaguri, no this does not fix my issue #813.

I confirm by putting log statements in the following places:

```
Within MainApplication.onCreate()
App.js right after all the imports
Within MusicService.onTaskRemoved()
Within MusicService.onDestroy()
```

The onCreate and App.js logs only happen once if I:

- launch the app
- start playing a track
- quit the app
- relaunch

This tells me that MainApplication, and the main JS instance are sticking around and never getting destroyed, as described in #813

**Comment by skahack** [comment]
Logs like this in my environment.
```
# This PR
TrackPlayerTest: onTaskRemoved
TrackPlayerTest: onTaskRemoved:stopSelf
TrackPlayerTest: destroy
TrackPlayerTest: onTaskRemoved
TrackPlayerTest: onTaskRemoved:stopSel
```

```
# Comment out manager.getPlayback().stop() and destroy().
TrackPlayerTest: onTaskRemoved
TrackPlayerTest: onTaskRemoved:stopSelf
TrackPlayerTest: onTaskRemoved
TrackPlayerTest: onTaskRemoved:stopSelf
TrackPlayerTest: onDestroy
TrackPlayerTest: destroy
```

You might not entered in this if-block Maybe.
https://github.com/react-native-kit/react-native-track-player/pull/779/files#diff-836c29352f6057f63a6cb9323634d8e9R130-R133

**Comment by curiousdustin** [comment]
Correct. This PR seems to function as intended.

However, I was answering the question about whether it solved #813. It does not. The application and JS state remain alive and in memory, even after the activity is destroyed.

**Comment by skahack** [comment]
I also had this problem #813 , but this PR has fixed it.
So if you can investigate why it doesn't enter in the block, it may be resolved.

**Comment by curiousdustin** [comment]
@skahack , oh interesting. hmmm.

So if you put logging in MainApplication.onCreate(), it will happen everytime? Even if you close the app while track is playing, and relaunch?

When I tested, I DID have that if-block, and it WAS running, it just doesn't end up killing the entire App/Process. 

I'm going to try again now.

**Comment by skahack** [comment]
> So if you put logging in MainApplication.onCreate(), it will happen everytime? Even if you close the app while track is playing, and relaunch?

```
# This PR
-> launch App
TrackPlayerTest: MainApplication.onCreate
-> kill App task
-> play track
TrackPlayerTest: onTaskRemoved
TrackPlayerTest: onTaskRemoved:stopSelf
TrackPlayerTest: destroy
TrackPlayerTest: onTaskRemoved
TrackPlayerTest: onTaskRemoved:stopSelf
-> launch App
-> kill App task
```

```
# Comment out manager.getPlayback().stop() and destroy().
-> launch App
TrackPlayerTest: MainApplication.onCreate
-> kill App task
-> play track
TrackPlayerTest: onTaskRemoved
TrackPlayerTest: onTaskRemoved:stopSelf
TrackPlayerTest: onTaskRemoved
TrackPlayerTest: onTaskRemoved:stopSelf
TrackPlayerTest: onDestroy
TrackPlayerTest: destroy
-> launch App
-> kill App task
TrackPlayerTest: onTaskRemoved
TrackPlayerTest: MainApplication.onCreate
```

**Comment by curiousdustin** [comment]
Hmm, I wonder what we are doing differently.

Here is what happens to me with this PR in place:

```
> Launch App

TrackPlayerTest: MainApplication.onCreate()
TrackPlayerTest: MainActivity.onCreate()
TrackPlayerTest: MainActivity.onCreate()
ReactNativeJS TrackPlayerTest App.js calling createStackNavigator()

> Play Track
> Recent Apps Menu, Remove App

TrackPlayerTest: MusicService.onTaskRemoved()
TrackPlayerTest: manager exists, calling manager.getPlayback().stop()
TrackPlayerTest: MusicService.onTaskRemoved calling destroy()
TrackPlayerTest: MusicService.onTaskRemoved calling stopSelf()
TrackPlayerTest: MusicService.onDestroy()

> Launch App

TrackPlayerTest: MainActivity.onCreate()
TrackPlayerTest: MainActivity.onCreate()

> DO NO Play Track
> Recent Apps Menu, Remove App
> Launch App

TrackPlayerTest: MainApplication.onCreate()
TrackPlayerTest: MainActivity.onCreate()
TrackPlayerTest: MainActivity.onCreate()
ReactNativeJS TrackPlayerTest App.js calling createStackNavigator()
```

As you can see, the MainApplication is only created again when launching the app after killing it before using the track player.

**Comment by curiousdustin** [comment]
Actually, your results appear the same. MainApplication.onCreate is not called again for you either when re-launching, is that correct?

**Comment by skahack** [comment]
Could you change this to `START_STICKY` ? I missed this PR (#714) being merged.
https://github.com/react-native-kit/react-native-track-player/blob/a53707d1c51515347d3db72820f0c20689c3ea26/android/src/main/java/com/guichaguri/trackplayer/service/MusicService.java#L112

**Comment by skahack** [comment]
I understand. This line caused the service to crash and terminate.
https://github.com/react-native-kit/react-native-track-player/pull/779/commits/a329770f71d22fc48e0037a63ef4a611f8eef183#diff-836c29352f6057f63a6cb9323634d8e9R130

**Comment by curiousdustin** [comment]
Released in [1.2.4](https://github.com/react-native-kit/react-native-track-player/releases/tag/v1.2.4)

**Pull request merged**